stingrayzboy/ebazaar
# Rails6 upgrade

Rails upgrade 5.2 to 6.1 :

Our goal here is to make sure that we use the latest version of ruby which at the time of writing is ruby 3.1.2.
our app currently works on ruby 2.7.0 and the rails version is 5.2.8.
Now rails 5 is not compatible with ruby 3 and also rails 6.0 is not compatible.
After researching on internet rails 6.1 is compatible.
The next step was to check what is the ruby version 2  which is compatible with rails 5.2 and 6.1.
So ruby 2.7.0 met this criteria.

The next step was to do a through check of our appln working smoothly.
Steps to do sanity check:
- Run rspec to check all specs are passing
- Ui testing all urls are action methods are working as expected.

After this we went into ruby gem , we looked for rails 6.1 entry and we add this entry and remove 5.2 entry. We also decided that this was the good time to upgrade puma as well.
After doing that we ran the cmd “bundle update rails” So this cmd might or might not give errors if it did give errors we might had to do make some changes in gemlock file and recreate it.

The next step was to run bundle install.
The next step was to do a through check of our appln working smoothly.
Steps to do sanity check:
- Run rspec to check all specs are passing
- Ui testing all urls are action methods are working as expected.

Then we decided to do a git commit.

Plz note currently till this step rails 6 although installed was still running this appln with 5.2 config.
So the next step was to update the app by running “rails app:update” . This cmd made a lot of changes in the interanl config code base we have. It also ask file by file basis which file to take and keep. We decided to choose the option ‘a’ which overwrote all the file that were required. We then use the git to look for the changes we made. Next step was to revert any unnecessary changes that were made ex: routes.rb. The changes also included some migration ,so we ran db:migrate.

The next step was to do a through check of our appln working smoothly.
Steps to do sanity check:
- Run rspec to check all specs are passing
- Ui testing all urls are action methods are working as expected.

Then we decided to do a git commit.

Now our appln is properly upgraded to rails 6.1. However, the appln was still using the legacy asset pipeline. 
Therefore we decided to use webpacker.
First add the entry of the latest version of webpacker in gem file.
Then run the cmd `rails webpacker:install`
This cmd runs and ask which version to be installed .we have found that version 5.4.3 to be the best fit and installed it. We then went through the code changes.
So webpacker generates a bunch of config js files of which the most important is the `application.js`. 

We made the following entries here
![Screenshot from 2023-02-10 11-27-04](https://user-images.githubusercontent.com/22894520/218020528-be7fdca8-22fc-441b-a3a3-783810a298c0.png)



Webpacker did not resolve them directly so we had to install them manually using “yarn”.
So we installed 
```bash
yarn add @rails/ujs
yarn add turbolinks
yarn add @rails/activestorage
yarn add bootstrap
yarn add @popperjs/core
yarn add noty
```

The next step was to move `app/assets/stylesheets/application.sccs` to  `app/javascript/stylesheets/application.sccs`
and imported `noty.scss`.
Then in application.html file we removed `<%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>`.
Changed the `<%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>` to `<%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>`.

Again did regression testing.
The next step was to do a through check of our appln working smoothly.
Steps to do sanity check:
- Run rspec to check all specs are passing
- Ui testing all urls are action methods are working as expected.

Now we don’t need to have bootstrap and jquery-rails gem. So went ahead and remove them and ran bundle install. 

We also decided to upgrade ruby v3.1.2 by 1st deleting the gemfile-lock .Then making changes .ruby-version with from 2.7.0 -> 3.1.2.
Also made changes in gemfile from 2.7.0 -> 3.1.2.
We ran bundle install .
We reconfigured overcommit.
We ran rspec and clean up the gems rspec warned about.



What works what does not works?

The base functionality of the app works fine but the steylesheet are breaking alittle bit bcoz the classes we have applied in the UI are of bootstrap 4 they needed to be coverted to bootstrap 5 equivalent.
